src/main/java/com/google/devtools/build/lib/skyframe/SkyframeExecutor.java

```diff
@@ -984,6 +984,37 @@ private static boolean isAnalysisPhaseActionLookupKey(SkyKey key) {
     return key instanceof ActionLookupKey && !(key instanceof ActionTemplateExpansionKey);
   }
 
+  /**
+   * A variant of {@link #DEFAULT_EVENT_FILTER_WITH_ACTIONS} that also handles Bzlmod events, which
+   * are only relevant for Bazel and its support for external repositories.
+   */
+  public static final EventFilter DEFAULT_EVENT_FILTER_WITH_ACTIONS_AND_EXTERNAL_REPOS =
+      new EventFilter() {
+        @Override
+        public boolean storeEvents() {
+          return true;
+        }
+
+        @Override
+        public boolean shouldPropagate(SkyKey depKey, SkyKey primaryKey) {
+          // Bzlmod events do not need to be propagated to analysis nodes. They are propagated to
+          // the node for the main repo mapping, which is unconditionally requested in a separate
+          // Skyframe evaluation for every build, thus ensuring correct incremental behavior.
+          if (isBzlmodKey(depKey)) {
+            return isBzlmodKey(primaryKey);
+          }
+          return DEFAULT_EVENT_FILTER_WITH_ACTIONS.shouldPropagate(depKey, primaryKey);
+        }
+      };
+
+  private static boolean isBzlmodKey(SkyKey key) {
+    // The return value of Class#getPackageName() and string literals are always interned.
+    // https://github.com/openjdk/jdk/blob/3e185c70feef3febf75c58a5d4d394a4b772105f/src/java.base/share/classes/java/lang/Class.java#L1258
+    return key.getClass().getPackageName() == "com.google.devtools.build.lib.bazel.bzlmod"
```

Comment:
this smells really fishy but I have to admit I don't really understand the `shouldPropagate` discussion, so take this with a grain of salt; but having code logic couple _so_ closely with code organization makes me very uncomfortable.

To confirm, is this needed for this PR? I was under the impression that the `shouldPropagate` stuff was related to the events generated during registry access that are sent to the Skyframe handlers. Do we also need it for some other purpose?

Reply:
Since this PR doesn't touch `BazelModuleResolutionEvent`, the change to `shouldPropagate` is needed to address the memory regression that retaining Bzlmod events caused (every Skyframe node comes with a `ValueWithMetadata` wrapper). The future PRs will get rid of `BazelModuleResolutionEvent`, but for that they need to introduce the download events, so that by itself wouldn't improve the situation.

I agree that the idea of checking the package is fishy, but I found it to be the least bad option I could come up with, others being:
1. Enumerate all Bzlmod `SkyKey` classes in some `Set`, but that would make membership testing slower and also require passing this Bazel/Bzlmod-specific logic into the executor.
2. Have Bzlmod `SkyKey`s implement a common interface, but pre-JDK 23 that could have [surprising performance implications](https://bugs.openjdk.org/browse/JDK-8180450) and we would still need to get this Bazel-specific type into `SkyframeExecutor`.
3. Do not use retained events in Bzlmod at all. We can get rid of `BazelModuleResolutionEvent` and try to build our own event mechanism for registry downloads. However, since we need to make sure that the events are correctly replayed on no-op builds with the lockfile removed, we would probably need to include them into the `SkyValue`s as in my original approach for https://github.com/bazelbuild/bazel/pull/21901. This would make it harder to decouple the registry download logic from the SkyFunctions using it.

No matter how brittle this looks, note that the relation defined by this check is lower and upper bounded by tests: If it matches too little or too many nodes, `BazelModuleResolutionTest` will fail due to analysis nodes having Bzlmod events attached or the main repo mapping node *not* having Bzlmod events attached.

I also don't love this approach, so I'm happy to change it out for anything else you prefer.

Reply:
I'd prefer most of the alternatives to the package string check. A `Set<SkyFunctionName>`, marker interface, `default boolean` method on `SkyKey`. Final decision to @Wyverald since he has more stake in bzlmod.

Reply:
> the change to `shouldPropagate` is needed to address the memory regression that retaining Bzlmod events caused (every Skyframe node comes with a `ValueWithMetadata` wrapper).

Is that regression introduced in this very PR, or some existing regression? If the latter, I'd say it's not so urgent as to require fixing right away. We could address it in a separate change, and have this PR focus solely on replacing (some) events with graph lookups.

> 3\. Do not use retained events in Bzlmod at all. We can get rid of `BazelModuleResolutionEvent` and try to build our own event mechanism for registry downloads. However, since we need to make sure that the events are correctly replayed on no-op builds with the lockfile removed, we would probably need to include them into the `SkyValue`s as in my original approach for [Store remote registry file hashes in the lockfile #21901](https://github.com/bazelbuild/bazel/pull/21901). 

I'm still in favor of this approach. It's minimally intrusive, and easier to follow. Skyframe is a big scary beast and the fact that we're just now discovering some big memory regression is testament to that; so I'd rather we avoid using event propagation if we don't need to. Key part "if we don't need to": to me it feels like propagating events across nodes is definitely valuable in some cases, but for just recording registry access it feels overkill.

Simple proposal: add `subscribe` and `unsubscribe` methods to `Registry`. Before accessing the registry, subscribe; during, record accesses; after (and in `finally`), unsubscribe. All this can be done within one SkyFunction.

> This would make it harder to decouple the registry download logic from the SkyFunctions using it.

I don't quite understand this part: what "coupling" are you referring to? And anyway, pretty much only two SkyFunctions need access to the registry -- ModuleFileFunction and RepoSpecFunction; it's not like it's a widely applicable concern.

Reply:
> Traverse the Skyframe graph down from `RepositoryMappingValue.key(RepositoryMapping.MAIN)` and collect the hashes from these values.

Or even just aggregate all the collected hashes in `BazelDepGraphValue`? Saves us the trouble of even having to perform graph lookups at all.